mecanum_drive_controller/test/test_mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,592 @@
+  }
+}
+
+// when ref_timeout = 0 expect reference_msg is accepted and command_interfaces
+// are calculated to non-nan values and reference_interfaces are set to nan
+TEST_F(MecanumDriveControllerTest, test_ref_timeout_zero_for_update)
+{
+  SetUpController();
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  auto reference = controller_->input_ref_.readFromNonRT();
+
+  // set command statically
+  joint_command_values_[1] = command_lin_x;
+
+  controller_->ref_timeout_ = rclcpp::Duration::from_seconds(0.0);
+  std::shared_ptr<ControllerReferenceMsg> msg = std::make_shared<ControllerReferenceMsg>();
+
+  msg->header.stamp = controller_->get_node()->now() - rclcpp::Duration::from_seconds(0.0);
+  msg->twist.linear.x = TEST_LINEAR_VELOCITY_X;
+  msg->twist.linear.y = TEST_LINEAR_VELOCITY_y;
+  msg->twist.linear.z = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.x = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.y = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.z = TEST_ANGULAR_VELOCITY_Z;
+  controller_->input_ref_.writeFromNonRT(msg);
+  const auto age_of_last_command =
+    controller_->get_node()->now() - (*(controller_->input_ref_.readFromNonRT()))->header.stamp;
+
+  fprintf(stderr, " age_of_last_command= %f \n", age_of_last_command);
+  fprintf(stderr, " controller_->ref_timeout_= %f \n", controller_->ref_timeout_);
+
+  ASSERT_FALSE(age_of_last_command <= controller_->ref_timeout_);
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  EXPECT_FALSE(std::isnan(joint_command_values_[1]));
+  EXPECT_NE(joint_command_values_[1], command_lin_x);
+  //  w0_vel = 1.0 / params_.kinematics.wheels_radius
+  // * (body_velocity_center_frame_.linear_x - body_velocity_center_frame_.linear_y
+  // - params_.kinematics.sum_of_robot_center_projection_on_X_Y_axis
+  // * body_velocity_center_frame_.angular_z);
+  //  joint_command_values_[1] = 1.0 / 0.5 * (1.5 - 0.0 - 1 * 0.0)
+  EXPECT_EQ(joint_command_values_[1], 3.0);
+  ASSERT_TRUE(std::isnan((*(controller_->input_ref_.readFromRT()))->twist.linear.x));
+}
+
+// when ref_timeout = 0 expect reference_callback() writes reference_msg to rt buffer
+// from nonrt thread
+TEST_F(MecanumDriveControllerTest, when_ref_timeout_zero_for_reference_callback_expect_reference_msg_in_rt_buffer)
```

Comment:
I think this test is duplicated.

Reply:
in test "when_reference_timeout_is_zero_expect_reference_msg_being_used_only_once" when update is called implicitly reference_callback is called. but this test explicitly shows the same by calling refernce_callback. and this was your idea to include it.

Reply:
Reference callback is not called implicitly when `update` is called. We are testing before update of the reference when timeout is set to 0, aren't we? So, I am still not sure what is added value of this test?

Reply:
true.Reference callback is not called implicitly when update is called, instead when a msg is published on the topic!


reference_callback() has ref_timeout check inside it
i remember we wrote this test to check this functionality . and another reason to keep track of any changes inside the ref_callback() in future?